Cache name resolutions

twm opened the pull request:
To avoid resolving the allowed hostnames once per test, cache the results in a shared dict. This should result in one set of resolutions per pytest run (or test worker, under pytest-xdist). I did this in a backward-compatible manner by introducing an optional parameter to the `socket_allow_hosts()` and `normalize_allowed_hosts()` functions as they appear to be public API, albeit undocumented.

I also refactored the plugin [to use Stash](https://docs.pytest.org/en/7.0.x/how-to/writing_hook_functions.html#plugin-stash), which was introduced in pytest 7.0.0, rather than set attributes on Config. If you want to keep support for `pytest>=6.2.5` I can back that bit out, I just thought it improved clarity.

miketheman commented on pytest_socket.py:
```diff
@@ -67,6 +69,18 @@ def socket_enabled(pytestconfig):
+    socket_force_enabled: bool
+    allow_unix_socket: bool
+    allow_hosts: typing.Union[str, typing.List[str], None]
+    resolution_cache: typing.Dict[str, typing.Set[str]]
```
I _think_ this can have a `default_factory` of an empty dict, so that when creating it later you don't have to pass an empty dict.

Refs: https://docs.python.org/3/library/dataclasses.html#dataclasses.field

twm commented on pytest_socket.py:
```diff
@@ -206,20 +232,29 @@ def resolve_hostnames(hostname: str) -> typing.Set[str]:
+def socket_allow_hosts(
+    allowed: typing.Union[str, typing.List[str], None] = None,
+    allow_unix_socket: bool = False,
+    resolution_cache: typing.Optional[typing.Dict[str, typing.List[str]]] = None,
```
:bug: This is proof that a test is required, because it should be passed down.

Now that you've approved the approach I'll add that! Thank you for the quick turn-around.

twm commented:
@miketheman I think I've addressed all your feedback. I settled on monkey patching `socket.getaddrinfo()` to count the number of calls to ensure the cache works.

Interestingly, it looks like pre-commit.ci's autocommit bypasses workflow approval. (Sorry about the many notifications...)

miketheman reviewed (approved):
@twm Thanks for putting in the effort!

twm commented:
Thank you for the review!
